EIPS/eip-7948.md

```diff
@@ -0,0 +1,139 @@
+
+Given that Ethereum’s database backend already supports sorted key-value storage and iteration, adding native support for a sorted KV store in the EVM is trivial from an implementation standpoint. Ethereum contracts would be able to interact with this underlying functionality by leveraging the existing database capabilities for iterating over keys in sorted order.
+
+Raw iterable key-value stores are fundamental building blocks of modern database systems. By introducing native support for a sorted KV store in the EVM, Ethereum contracts would be empowered to build sophisticated data management and querying libraries. These libraries could take full advantage of the underlying database features to support complex use cases. This would open up new possibilities for developers and increase the flexibility and efficiency of smart contract development.
+
+### Use Cases
+
+- **Ranking Systems**: Automatically maintain a leaderboard where scores or ranks are updated in real-time and can be iterated in sorted order.
+- **Auctions**: Maintain and iterate over auction bids in ascending or descending order.
+- **Voting Systems**: Efficiently track and iterate over votes or other metrics in sorted order.
+- **Order books**: Manage and iterate over buy and sell orders in price order for decentralized exchanges or markets.
+
+### Potential Alternatives
+
+- **Off-Chain Solutions**: Currently, many developers use off-chain solutions, such as oracles or indexing services, to maintain sorted data. However, these solutions increase complexity and reliance on external infrastructure.
+- **Custom Sorting Logic**: Developers could manually implement sorting within contracts, but this requires complex logic and would be costly in terms of gas fees. The native support for sorted KV stores would streamline this process.
+
+## Backwards Compatibility
+
+The new functionality would not break existing smart contracts but would add a new method of managing key-value pairs. It is fully compatible with the existing Ethereum Virtual Machine (EVM) and can coexist with current storage models such as mappings.
+
+## Security Considerations
+
+Since the values in Indexed Storage can be of arbitrary size, a key associated with a large value could result in unexpectedly high gas costs when read. This could lead to the caller unintentionally paying excessive gas fees. To mitigate this risk, one solution would be to limit values to 32 bytes, similar to the fixed-size keys, ensuring more predictable gas costs and preventing excessive storage usage.
```

Comment:
on system level, every storage is read as 32 bytes, its upto compilers to constuct high level data structures.

so for what you are proposing is the security consideration is: that since keys are stored in a sequence it will lead to an unbalanaced state tree with higher costs to access those keys